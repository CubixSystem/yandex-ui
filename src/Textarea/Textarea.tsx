import React, { FC, Ref, ReactNode, MouseEventHandler } from 'react';
import { useComponentRegistry } from '@bem-react/di';
import { cn } from '@bem-react/classname';

import { RenderOverride, useRenderOverride } from '../lib/render-override';
import { useTextFieldHint } from '../useTextFieldHint';
import { IWithControlProps, withControl } from '../withControl/withControl';
import { IWithControlProps as IWithControlPropsDesktop } from '../withControl/withControl@desktop';
import { ITextareaControlProps } from './Control/Textarea-Control';
import { ITextareaRegistry } from './Textarea.registry';
import './Textarea.css';

export interface ITextareaProps
    extends IWithControlProps<HTMLTextAreaElement>,
        // https://st.yandex-team.ru/ISL-6012
        IWithControlPropsDesktop<HTMLTextAreaElement>,
        ITextareaControlProps {
    /**
     * Событие, которое вызывается при нажатии на компонент
     */
    onClick?: MouseEventHandler<HTMLElement>;

    /**
     * Событие по своему действию похоже на `onClick` и возникает в момент нажатия на кнопку мыши.
     * `onClick` в каком-то смысле является комбинацией событий `onMouseDown` и `onMouseUp`
     */
    onMouseDown?: MouseEventHandler<HTMLElement>;

    /**
     * Событие по своему действию противоположно событию `onMouseDown` и происходит при отпускании кнопки мыши.
     * Курсор должен находится в пределах HTML-элемента, к которому добавлен атрибут `onmouseup`
     */
    onMouseUp?: MouseEventHandler<HTMLElement>;

    /**
     * Обработчик события `onMouseLeave`
     */
    onMouseLeave?: MouseEventHandler<HTMLElement>;

    /**
     * Обработчик события `onMouseEnter`
     */
    onMouseEnter?: MouseEventHandler<HTMLElement>;

    /**
     * Состояние фокуса на компоненте
     */
    focused?: boolean;

    /**
     * Состояние нажатия на компоненте
     */
    pressed?: boolean;

    /**
     * Неактивное состояние компонента.
     * Состояние, при котором компонент отображается, но недоступен для действий пользователя
     */
    disabled?: boolean;

    /**
     * Состояние, которое возникает при наведении на компонент курсором
     */
    hovered?: boolean;

    /**
     * Дополнительный контент после контрола.
     */
    addonAfter?: ReactNode;

    /**
     * Дополнительный контент перед контролом.
     */
    addonBefore?: ReactNode;

    /**
     * Ссылка на корневой DOM элемент компонента.
     */
    innerRef?: Ref<HTMLSpanElement>;

    /**
     * Ссылка на враппер DOM-элемента нативного инпута.
     */
    wrapRef?: Ref<HTMLSpanElement>;

    /**
     * Текст-подсказка, появляющаяся после компонента.
     * Может иметь различное визуальное оформление в зависимости от свойства `state`.
     */
    hint?: string;

    /**
     * Визуальное состояние компонента.
     * Может использоваться при проверке формы на корректность.
     */
    state?: 'error';

    /**
     * Всплывающая подсказка
     */
    title?: string;

    /**
     * Переопределяет компонент `Control`
     */
    renderControl?: RenderOverride<ITextareaControlProps>;
}

export const cnTextarea = cn('Textarea');

/**
 * Многострочное текстовое поле.
 * @param {ITextareaProps} props
 */
const TextareaBase: FC<ITextareaProps> = ({
    addonAfter,
    addonBefore,
    disabled,
    focused,
    hovered,
    onMouseEnter,
    onMouseLeave,
    innerRef,
    wrapRef,
    className: textareaClassName,
    // Извлекаем свойства, т.к. они не нужны на DOM узле
    // FIXME: https://github.com/bem/bem-react/issues/381
    pressed: _pressed,
    // @ts-ignore
    view: _view,
    // @ts-ignore
    size: _size,
    // @ts-ignore
    theme: _theme,
    hint: htmlHint,
    state,
    title,
    renderControl,
    ...props
}) => {
    const className = cnTextarea(
        {
            hovered,
            focused,
            disabled,
            state,
        },
        [textareaClassName],
    );

    const { Wrap, Control: ControlOriginal, Box, Hint } = useComponentRegistry<ITextareaRegistry>(cnTextarea());
    const Control = useRenderOverride(ControlOriginal, renderControl);

    const { hint, hintProps } = useTextFieldHint({ hint: htmlHint });

    return (
        <span
            ref={innerRef}
            className={className}
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
            title={title}
        >
            {addonBefore}
            <Wrap innerRef={wrapRef}>
                <Control {...props} aria-invalid={state === 'error'} disabled={disabled} />
                <Box />
            </Wrap>
            {addonAfter}
            {hint && <Hint {...hintProps}>{hint}</Hint>}
        </span>
    );
};

TextareaBase.displayName = cnTextarea();

export const Textarea = withControl(TextareaBase);
