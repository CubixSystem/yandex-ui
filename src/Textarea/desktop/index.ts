export * from '../Textarea@desktop';
export * from '../_autoResize/Textarea_autoResize@desktop';
export * from '../_hasClear/Textarea_hasClear@desktop';
export * from '../_size/Textarea_size_m';
export * from '../_size/Textarea_size_s';
export * from '../_theme/Textarea_theme_normal@desktop';
export * from '../_view/Textarea_view_default@desktop';
