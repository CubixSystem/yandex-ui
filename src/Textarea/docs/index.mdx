---
id: Textarea
path: /components/textarea

title: Textarea
description: Многострочное текстовое поле.

links:
 - label: source
   url: https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/Textarea
 - label: startrek
   url: https://search.yandex-team.ru/stsearch?text=Textarea.ts&amp;facet.queue=ISL&amp;facet.type=bug&amp;facet.status=128
 - label: figma
   url: https://www.figma.com/file/CaoHFEO0w6PaE8cVaYFDoq8i/LEGO?node-id=249%3A1049
 - label: storybook
   url: https://lego.yandex-team.ru/components/lego/latest/?storybook=%2Fstory%2Fcontrols-textinput-desktop--playground

tabs:
 - Examples
 - Code
---
