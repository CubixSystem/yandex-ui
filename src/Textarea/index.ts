export * from './Textarea';
export * from './_autoResize/Textarea_autoResize';
export * from './_hasClear/Textarea_hasClear';
export * from './_size/Textarea_size_m';
export * from './_size/Textarea_size_s';
export * from './_theme/Textarea_theme_normal';
export * from './_view/Textarea_view_default';
