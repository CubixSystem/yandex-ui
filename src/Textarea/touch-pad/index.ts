export * from '../Textarea@touch-pad';
export * from '../_autoResize/Textarea_autoResize@touch-pad';
export * from '../_hasClear/Textarea_hasClear@touch-pad';
export * from '../_size/Textarea_size_m';
export * from '../_size/Textarea_size_s';
export * from '../_theme/Textarea_theme_normal';
export * from '../_view/Textarea_view_default';
