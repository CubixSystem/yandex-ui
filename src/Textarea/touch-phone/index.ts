export * from '../Textarea@touch-phone';
export * from '../_autoResize/Textarea_autoResize@touch-phone';
export * from '../_hasClear/Textarea_hasClear@touch-phone';
export * from '../_size/Textarea_size_m';
export * from '../_size/Textarea_size_s';
export * from '../_theme/Textarea_theme_normal';
export * from '../_view/Textarea_view_default';
