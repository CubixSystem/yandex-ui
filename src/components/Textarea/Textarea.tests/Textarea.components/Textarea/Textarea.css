.Textarea_size_m {
    height: 64px;
}

.Textarea_size_s {
    height: 56px;
}
