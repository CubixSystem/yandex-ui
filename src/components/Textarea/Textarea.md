# Textarea

<a href="https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/components/Textarea" target='_blank'><img src='https://badger.yandex-team.ru/custom/[Исходники]/[Github
][green]/badge.svg' /></a> <a href="https://search.yandex-team.ru/stsearch?text=Textarea.ts&amp;facet.queue=ISL&amp;facet.type=bug&amp;facet.status=128" target='_blank'><img src='https://badger.yandex-team.ru/custom/[Известные проблемы]/[Startrek][blue]/badge.svg' /></a>

<!-- description:start -->
Многострочное текстовое поле.
<!-- description:end -->

## Пример использования

Конфигурация темы на уровне проекта:

```ts
// src/lib/theme.ts
import { configureRootTheme } from '@yandex-lego/components/Theme'
import { theme } from '@yandex-lego/components/Theme/presets/default'

configureRootTheme({ theme })
```

Использование с нужным набором модификаторов:

```ts
// src/App.ts
import React, { useState } from 'react'
import { compose } from '@bem-react/core'
import {
  Textarea as TextareaDesktop,
  withViewDefault,
  withSizeM,
} from '@yandex-lego/components/Textarea/desktop'

// Композиция из различных модификаторов
const Textarea = compose(withViewDefault, withSizeM)(TextareaDesktop)

const App = () => {
  const [value, setValue] = useState('')

  return (
    <Textarea
      size="m"
      view="default"
      value={value}
      onChange={(event) => setValue(event.target.value)}
    />
  )
}
```

Использование с полным набором модификаторов:

```ts
// src/App.ts
import React, { useState } from 'react'
import { Textarea } from '@yandex-lego/components/Textarea/desktop/bundle'

const App = () => {
  const [value, setValue] = useState('')

  return (
    <Textarea
      size="m"
      view="default"
      value={value}
      onChange={(event) => setValue(event.target.value)}
    />
  )
}
```

## Примеры

### Вид поля

Чтобы изменить вид текстового поля, установите свойство `view` в значение `"default"`.

{{%story::desktop:controls-textarea-desktop--view%}}

### Стилевое оформление

Чтобы изменить стилевое оформление текстового поля, установите свойство `theme` в значение `"normal"`.

{{%story::desktop:controls-textarea-desktop--theme%}}

### Размер поля

Чтобы изменить размер текстового поля, установите свойство `size` в одно из следующих значений: `"m"`, `"s"`.

{{%story::desktop:controls-textarea-desktop--size%}}

### Крестик для очистки

Чтобы добавить крестик для очистки текстового поля, установите свойство `hasClear`.
Для появления крестика установите не пустое `value` в компоненте (не работает с `defaultValue`).

{{%story::desktop:controls-textarea-desktop--has-clear%}}

| Свойство      | Тип                                                        | Описание                                      |
| ------------- | ---------------------------------------------------------- | --------------------------------------------- |
| hasClear?     | `false \| true`                                            | Наличие крестика для очистки текстового поля. |
| onClearClick? | `(event: MouseEvent<HTMLSpanElement, MouseEvent>) => void` | Обработчик клика по крестику.                 |

### Состояние ошибки

Чтобы задать состояние ошибки нужно передать свойство `state` со значением `error`.
Чтобы установить текст ошибки нужно передать значение в свойство `hint`.

{{%story::desktop:controls-textarea-desktop--state%}}

## Свойства

<!-- props:start -->
| Свойство      | Тип                                                                                                                                                                                                                                                               | Описание                                                                                                                                                                                           |
| ------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| onBlur?       | `(event: FocusEvent<HTMLElement>) => void`                                                                                                                                                                                                                        | Событие, которое вызывается при потере фокуса компонентом. Например, при клике на другом месте экрана                                                                                              |
| onClick?      | `(event: MouseEvent<HTMLElement, MouseEvent>) => void`                                                                                                                                                                                                            | Событие, которое вызывается при нажатии на компонент                                                                                                                                               |
| onKeyDown?    | `(event: KeyboardEvent<HTMLElement>) => void`                                                                                                                                                                                                                     | Событие, которое вызывается при нажатии клавиш клавиатуры                                                                                                                                          |
| onFocus?      | `(event: FocusEvent<HTMLElement>) => void`                                                                                                                                                                                                                        | Событие, которое возникает при получении компонентом фокуса                                                                                                                                        |
| onMouseDown?  | `(event: MouseEvent<HTMLElement, MouseEvent>) => void`                                                                                                                                                                                                            | Событие по своему действию похоже на `onClick` и возникает в момент нажатия на кнопку мыши. `onClick` в каком-то смысле является комбинацией событий `onMouseDown` и `onMouseUp`                   |
| onMouseUp?    | `(event: MouseEvent<HTMLElement, MouseEvent>) => void`                                                                                                                                                                                                            | Событие по своему действию противоположно событию `onMouseDown` и происходит при отпускании кнопки мыши. Курсор должен находится в пределах HTML-элемента, к которому добавлен атрибут `onmouseup` |
| onMouseLeave? | `(event: MouseEvent<HTMLElement, MouseEvent>) => void`                                                                                                                                                                                                            | Обработчик события `onMouseLeave`                                                                                                                                                                  |
| onMouseEnter? | `(event: MouseEvent<HTMLElement, MouseEvent>) => void`                                                                                                                                                                                                            | Обработчик события `onMouseEnter`                                                                                                                                                                  |
| focused?      | `false \| true`                                                                                                                                                                                                                                                   | Состояние фокуса на компоненте                                                                                                                                                                     |
| pressed?      | `false \| true`                                                                                                                                                                                                                                                   | Состояние нажатия на компоненте                                                                                                                                                                    |
| disabled?     | `false \| true`                                                                                                                                                                                                                                                   | Неактивное состояние компонента. Состояние, при котором компонент отображается, но недоступен для действий пользователя                                                                            |
| hovered?      | `false \| true`                                                                                                                                                                                                                                                   | Состояние, которое возникает при наведении на компонент курсором                                                                                                                                   |
| addonAfter?   | `string \| number \| false \| true \| {} \| ReactElement<any, string \| ((props: any) => ReactElement<any, string \| ... \| (new (props: any) => Component<any, any, any>)>) \| (new (props: any) => Component<any, any, any>)> \| ReactNodeArray \| ReactPortal` | Дополнительный контент после контрола.                                                                                                                                                             |
| addonBefore?  | `string \| number \| false \| true \| {} \| ReactElement<any, string \| ((props: any) => ReactElement<any, string \| ... \| (new (props: any) => Component<any, any, any>)>) \| (new (props: any) => Component<any, any, any>)> \| ReactNodeArray \| ReactPortal` | Дополнительный контент перед контролом.                                                                                                                                                            |
| innerRef?     | `(instance: HTMLSpanElement) => void \| RefObject<HTMLSpanElement>`                                                                                                                                                                                               | Ссылка на корневой DOM элемент компонента.                                                                                                                                                         |
| hint?         | `string`                                                                                                                                                                                                                                                          | Текст-подсказка, появляющаяся после компонента. Может иметь различное визуальное оформление в зависимости от свойства `state`.                                                                     |
| state?        | `"error"`                                                                                                                                                                                                                                                         | Визуальное состояние компонента. Может использоваться при проверке формы на корректность.                                                                                                          |
| controlRef?   | `(instance: T) => void \| RefObject<T>`                                                                                                                                                                                                                           | Ссылка на дом-ноду нативного контрола.                                                                                                                                                             |
| onChange?     | `(event: ChangeEvent<HTMLTextAreaElement>) => void`                                                                                                                                                                                                               | Обработчик изменения значения                                                                                                                                                                      |
| autoFocus?    | `false \| true`                                                                                                                                                                                                                                                   | HTML-атрибут `autofocus`                                                                                                                                                                           |
| id?           | `string`                                                                                                                                                                                                                                                          | Уникальный id компонента                                                                                                                                                                           |
| name?         | `string`                                                                                                                                                                                                                                                          | HTML-атрибут `name`                                                                                                                                                                                |
| placeholder?  | `string`                                                                                                                                                                                                                                                          | Плейсхолдер                                                                                                                                                                                        |
| className?    | `string`                                                                                                                                                                                                                                                          | Дополнительный класс                                                                                                                                                                               |
| cols?         | `number`                                                                                                                                                                                                                                                          | HTML-атрибут `cols`                                                                                                                                                                                |
| rows?         | `number`                                                                                                                                                                                                                                                          | HTML-атрибут `rows`                                                                                                                                                                                |
| value?        | `string`                                                                                                                                                                                                                                                          | Значение контрола                                                                                                                                                                                  |
| defaultValue? | `string`                                                                                                                                                                                                                                                          | Значение по умолчанию контрола                                                                                                                                                                     |
<!-- props:end -->
