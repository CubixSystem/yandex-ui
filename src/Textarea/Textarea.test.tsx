import React, { createRef, ComponentType } from 'react';
import { mount, render } from 'enzyme';
import { ExtractProps } from '@bem-react/core';

import { Textarea as TextareaCommon } from './Textarea';
import { Textarea as TextareaDesktop } from './Textarea@desktop';
import { Textarea as TextareaTouchPad } from './Textarea@touch-pad';
import { Textarea as TextareaTouchPhone } from './Textarea@touch-phone';
import { serverEnvironmentSetup, delay } from '../internal/utils';

const platforms = [['desktop', TextareaDesktop], ['touch-pad', TextareaTouchPad], ['touch-phone', TextareaTouchPhone]];

type TextareaProps = ExtractProps<typeof TextareaDesktop>;

describe.each<any>(platforms)('Textarea@%s', (_platform, Textarea: ComponentType<TextareaProps>) => {
    describe('server environment', () => {
        serverEnvironmentSetup();

        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(render(<Textarea className="mix" />)).toMatchSnapshot();
        });
    });

    describe('client environment', () => {
        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(render(<Textarea className="mix" />)).toMatchSnapshot();
        });

        test('должен быть установлен корректный displayName', () => {
            //Так написано, потому что компонент сразу поставляется в обертке.
            expect(TextareaCommon.displayName).toBe('WithControl(Textarea)');
        });

        test('должен устанавливать ссылку на корневой DOM элемент', async () => {
            const innerRef = createRef<HTMLElement>();
            mount(<Textarea innerRef={innerRef} />);
            await delay(100);
            expect(innerRef.current).not.toBe(null);
        });

        test('должен устанавливать ссылку на нативный DOM инпут', async () => {
            const controlRef = createRef<HTMLTextAreaElement>();
            mount(<Textarea controlRef={controlRef} />);
            await delay(100);
            expect(controlRef.current).not.toBe(null);
        });

        test('должен вызывать событие onChange с event при вводе текста в поле', () => {
            const onChange = jest.fn();
            const wrapper = mount(<Textarea onChange={onChange} />);
            wrapper.find('textarea').simulate('change', { target: { value: 'next-value' } });
            expect(onChange).toHaveBeenCalledTimes(1);
            expect(onChange.mock.calls[0][0].target.value).toBe('next-value');
        });

        test('должен обновлять value в textarea при обновлении через props', () => {
            const wrapper = mount(<Textarea value="" onChange={() => null} />);
            wrapper.setProps({ value: 'next' });
            expect(wrapper.find('textarea').prop('value')).toBe('next');
            wrapper.setProps({ value: 'next-value' });
            expect(wrapper.find('textarea').prop('value')).toBe('next-value');
        });

        test('должен рендерить контент в addonAfter и addonBefore', () => {
            const wrapper = mount(
                <Textarea addonAfter={<i className="after" />} addonBefore={<i className="before" />} />,
            );
            expect(wrapper.find('.after')).toHaveLength(1);
            expect(wrapper.find('.before')).toHaveLength(1);
        });

        test('должен добавлять и удалять атрибут disabled при обновлении через props', () => {
            const wrapper = mount(<Textarea />);
            wrapper.setProps({ disabled: true });
            expect(wrapper.find('textarea').prop('disabled')).toBe(true);
            wrapper.setProps({ disabled: false });
            expect(wrapper.find('textarea').prop('disabled')).toBe(false);
        });

        test('должен рендерить/удалять элемент Hint при наличии свойства', () => {
            const wrapper = mount(<Textarea hint="error" />);
            expect(wrapper.find('.Textarea-Hint')).toHaveLength(1);
            wrapper.setProps({ hint: '' });
            wrapper.find('.Textarea-Hint').simulate('animationend');
            expect(wrapper.find('.Textarea-Hint')).toHaveLength(0);
        });

        test('renderControl рисует div и туда прорастают свойства', () => {
            const wrapper = mount(
                <Textarea renderControl={(props:any) => <div className="Textarea-Control" {...props} />} name="textareaName" />,
            );

            expect(wrapper.find('.Textarea-Control').getDOMNode().tagName).toBe('DIV');
            expect(wrapper.find('.Textarea-Control').prop('name')).toBe('textareaName');
        });
    });
});
