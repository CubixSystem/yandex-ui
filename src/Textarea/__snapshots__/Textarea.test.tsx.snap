// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea@desktop client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textarea mix"
>
  <span
    class="Textarea-Wrap"
  >
    <textarea
      aria-invalid="false"
      class="Textarea-Control"
    />
    <span
      class="Textarea-Box"
    />
  </span>
</span>
`;

exports[`Textarea@desktop server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textarea mix"
>
  <span
    class="Textarea-Wrap"
  >
    <textarea
      aria-invalid="false"
      class="Textarea-Control"
    />
    <span
      class="Textarea-Box"
    />
  </span>
</span>
`;

exports[`Textarea@touch-pad client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textarea mix"
>
  <span
    class="Textarea-Wrap"
  >
    <textarea
      aria-invalid="false"
      class="Textarea-Control"
    />
    <span
      class="Textarea-Box"
    />
  </span>
</span>
`;

exports[`Textarea@touch-pad server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textarea mix"
>
  <span
    class="Textarea-Wrap"
  >
    <textarea
      aria-invalid="false"
      class="Textarea-Control"
    />
    <span
      class="Textarea-Box"
    />
  </span>
</span>
`;

exports[`Textarea@touch-phone client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textarea mix"
>
  <span
    class="Textarea-Wrap"
  >
    <textarea
      aria-invalid="false"
      class="Textarea-Control"
    />
    <span
      class="Textarea-Box"
    />
  </span>
</span>
`;

exports[`Textarea@touch-phone server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textarea mix"
>
  <span
    class="Textarea-Wrap"
  >
    <textarea
      aria-invalid="false"
      class="Textarea-Control"
    />
    <span
      class="Textarea-Box"
    />
  </span>
</span>
`;
